src/message_bus.zig

```diff
@@ -60,73 +233,419 @@ pub const MessageBus = struct {
+    /// This is guaranteed to be valid only while peer is unknown, client, or replica.
+    /// It will be reset to -1 during the shutdown process and is always -1 if the
+    /// connection is unused (i.e. peer == .none). We use -1 instead of undefined here
+    /// for safety to ensure an error if the invalid value is ever used, instead of
+    /// potentially preforming an action on an active fd.
+    fd: os.socket_t = -1,
+
+    /// This completion is used for all recv operations.
+    /// It is also used for the initial connect when establishing a replica connection.
+    recv_completion: IO.Completion = undefined,
+    /// True exactly when the recv_completion has been submitted to the IO abstraction
+    /// but the callback has not yet been run.
+    recv_in_progress: bool = false,
+    /// Number of bytes of the current header/message that have already been recieved.
+    recv_progress: usize = 0,
+    incoming_header: Header = undefined,
+    incoming_message: *Message = undefined,
+
+    /// This completion is used for all send operations.
+    send_completion: IO.Completion = undefined,
+    /// True exactly when the send_completion has been submitted to the IO abstraction
+    /// but the callback has not yet been run.
+    send_in_progress: bool = false,
+    /// Number of bytes of the current message that have already been sent.
+    send_progress: usize = 0,
+    /// The queue of messages to send to the client.
+    /// Empty unless peer == .client
+    send_queue: FIFO(Envelope) = .{},
+
+    /// Attempt to connect to a replica. Failure is silent and returns the
+    /// connection to an unused state.
+    pub fn connect_to_replica(self: *Connection, replica: u16) void {
+        assert(self.peer == .none);
+        assert(self.fd == -1);
+
+        const bus = self.message_bus;
+        const server_addr = bus.configuration[bus.server.replica];
+        self.fd = os.socket(server_addr.any.family, os.SOCK_STREAM | os.SOCK_CLOEXEC, 0) catch return;
+
+        self.peer = .{ .replica = replica };
+        bus.replicas[replica] = self;
+        bus.replicas_connected += 1;
+
+        self.recv_in_progress = true;
+        bus.io.connect(
+            *Connection,
+            self,
+            on_connect,
+            // We need to use the recv_completion here
+            &self.recv_completion,
+            self.fd,
+            // TODO: is this actually safe to pass by value?
+            // Should IO just take an std.net.Address or a pointer?
+            bus.configuration[replica].any,
+            bus.configuration[replica].getOsSockLen(),
+        );
+    }
+
+    fn on_connect(self: *Connection, completion: *IO.Completion, result: IO.ConnectError!void) void {
+        self.recv_in_progress = false;
+        result catch |err| {
+            log.err("error connecting to {}: {}", .{ self.peer, err });
+            self.shutdown();
+            return;
+        };
+        log.debug("connected to {}", .{self.peer});
```

Comment:
> I've been wanting to ask you for some time how we can reduce some of the repetitive boilerplate in src/vr.zig around .{self.replica} when logging. e.g. Add Replica.debug(), info() etc. It's not critical but would be nice to clean up.

If only running a single replica in the process, this would be as simple as defining a custom log function to add the replica index before logging.

If we want to support this logging format for multiple replicas in the same process, it would probably make the most sense to define a `Replica.log()` function which adds the replica id.